src/sentry/static/sentry/app/locale.tsx

```diff
@@ -0,0 +1,353 @@
+import React from 'react';
+import Jed from 'jed';
+import {sprintf} from 'sprintf-js';
+import {isString, isArray, isObject} from 'lodash';
+import {css} from 'react-emotion';
+
+import {getTranslations} from 'app/translations';
+
+const markerCss = css`
+  background: #ff801790;
+  outline: 2px solid #ff801790;
+`;
+
+const sessionStorage = window.sessionStorage;
+const LOCALE_DEBUG = sessionStorage && sessionStorage.getItem('localeDebug') === '1';
+
+export function setLocaleDebug(value: boolean) {
+  sessionStorage.setItem('localeDebug', value ? '1' : '0');
+  // eslint-disable-next-line no-console
+  console.log(`Locale debug is: ${value ? 'on' : 'off'}. Reload page to apply changes!`);
+}
+
+export function toggleLocaleDebug() {
+  const currentValue = sessionStorage.getItem('localeDebug');
+  setLocaleDebug(currentValue !== '1');
+  window.location.reload();
+}
+
+/**
+ * Global Jed locale object loaded with translations via setLocale
+ */
+let i18n: any = null;
+
+/**
+ * Set the current application locale.
+ *
+ * NOTE: This MUST be called early in the application before calls to any
+ * translation functions, as this mutates a singleton translation object used
+ * to lookup translations at runtime.
+ */
+export function setLocale(locale: string) {
+  const translations = getTranslations(locale);
+  i18n = new Jed({
+    domain: 'sentry',
+    missing_key_callback: () => {},
+    locale_data: {
+      sentry: translations,
+    },
+  });
+}
+
+setLocale('en');
+
+type FormatArg = ComponentMap | React.ReactNode;
+
+/**
+ * printf style string formatting which render as react nodes.
+ */
+function formatForReact(formatString: string, args: FormatArg[]) {
+  const nodes: React.ReactNodeArray = [];
```

Comment:
jw, is this type different than React.ReactNode[]?

Reply:
It's the same.